src/components/node-list/node-list-groups.js

```diff
@@ -1,122 +1,84 @@
 
   // Collapse/expand node group
-  const onToggleCollapsed = typeID => {
+  const onToggleGroupCollapsed = typeID => {
     const groupsCollapsed = Object.assign({}, collapsed, {
       [typeID]: !collapsed[typeID]
     });
     setCollapsed(groupsCollapsed);
     saveState({ groupsCollapsed });
   };
 
-  // Toggle node selection depending on whether it's already selected
-  const handleNodeSelection = nodeID => {
-    if (nodeSelected[nodeID]) {
-      onToggleNodeClicked(null);
-    } else {
-      onToggleNodeClicked(nodeID);
-    }
-  };
-
-  const renderTypeGroup = type => {
-    if (!nodes[type.id]) {
-      return null;
-    }
-    return (
-      <NodeListGroup
-        key={type.id}
-        onToggleCollapsed={onToggleCollapsed}
-        type={type}
-        collapsed={collapsed[type.id]}>
-        <ul className="pipeline-nodelist pipeline-nodelist--nested">
-          {nodes[type.id].map(node => (
-            <li key={node.id}>
-              <NodeListRow
-                active={nodeActive[node.id]}
-                checked={!node.disabled_node}
-                disabled={node.disabled_tag || node.disabled_type}
-                id={node.id}
-                label={node.highlightedLabel}
-                name={node.name}
-                onClick={() => handleNodeSelection(node.id)}
-                onMouseEnter={() => onToggleNodeHovered(node.id)}
-                onMouseLeave={() => onToggleNodeHovered(null)}
-                onChange={e => {
-                  onToggleNodesDisabled([node.id], !e.target.checked);
-                }}
-                selected={nodeSelected[node.id]}
-                type={node.type}
-              />
-            </li>
-          ))}
-        </ul>
-      </NodeListGroup>
-    );
-  };
-
-  return (
-    <Flipper flipKey={collapsed}>
-      <ul className="pipeline-nodelist">{types.map(renderTypeGroup)}</ul>
-    </Flipper>
-  );
+  return sections.map(section => (
+    <nav className="pipeline-nodelist-section kedro" key={section.name}>
+      <h4 className="pipeline-nodelist-section__title">{section.name}</h4>
+      <ul className="pipeline-nodelist">
+        {section.types.map(typeId => {
+          const group = groups[typeId];
+          return (
+            <NodeListGroup
+              key={group.id}
+              id={group.id}
+              name={group.name}
+              kind={group.kind}
+              checked={group.checked}
+              childCount={group.count}
+              allUnset={group.allUnset}
+              visibleIcon={group.visibleIcon}
+              invisibleIcon={group.invisibleIcon}
+              collapsed={Boolean(searchValue) ? false : collapsed[group.id]}
+              onToggleCollapsed={onToggleGroupCollapsed}
+              onToggleChecked={onToggleGroupChecked}>
+              <ul className="pipeline-nodelist pipeline-nodelist--nested">
+                {(items[group.id] || []).map(item => (
+                  <li key={item.id}>
```

Comment:
I noticed that this results in the following HTML structure:
![image](https://user-images.githubusercontent.com/1155816/95447770-29ec7100-095a-11eb-8fea-e92f52dc95fe.png)

Why not change the parent element in `NodeListRow` from a `<div>` to a `<li>`, and get rid of this parent element, to avoid having an unnecessary container dom node?

Reply:
I'm not sure that's a good idea personally: creating components that use `<li>`as their root limits them to only being used inside list elements. I've already noticed the same problem with the `IconButton` component, which has no real reason to be a list element. As this is already pre-existing in the codebase too, my preference is to avoid tightly coupling components to their parent's markup.

Reply:
Fragments won't work here as classnames are needed on the container unfortunately. I'm with you here though that this is a good idea to reduce overhead since there can be a lot of rows.

So Iistead I've added a prop `container` that allows optionally passing the tag name to `NodeListRow` and something similar for the children container of `NodeListGroup` (which is needed for the open/close functionality).

This approach then allowed those extra containers to be removed, while still maintaining separation of concerns and the flexibility of these components as is currently needed.